File: app/src/main/java/com/example/opengl/ResourceDepository.java
package com.example.opengl;

import com.example.opengl.devourer.DevourerNode;

import java.util.List;
import java.util.Timer;
import java.util.TimerTask;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;

public class ResourceDepository {
    private final BlockingQueue<Resource> resourceQueue = new LinkedBlockingQueue<>();
    private final int dispatchPeriod;
    private Timer timer;

    public ResourceDepository(int dispatchPeriod) {
        this.dispatchPeriod = dispatchPeriod;
    }

    public Resource takeResource() {
        try {
            return resourceQueue.take();
        } catch (InterruptedException e) {
            return null;
        }
    }

    public void addResource(Resource resource) {
        resourceQueue.add(resource);
    }

    public int getSize() {
        return resourceQueue.size();
    }

    public int getDispatchPeriod() {
        return dispatchPeriod;
    }

    public void startResourceMining(DevourerNode startDevourerNode, Position startPosition, List<Resource> movingResources) {
        if (getSize() > 0) {
            timer = new Timer();
            timer.schedule(new TimerTask() {
                public void run() {
                    Resource resource = takeResource();
                    if (resource == null) {
                        timer.cancel();
                        return;
                    }
                    resource.startResource(startDevourerNode, startPosition);
                    movingResources.add(resource);
                }
            }, 0, dispatchPeriod);
        }
    }


}

File: app/src/main/java/com/example/opengl/ResourceDepository.kt
package com.example.opengl

import com.example.opengl.devourer.DevourerStructure
import java.util.Timer
import java.util.TimerTask
import java.util.concurrent.BlockingQueue
import java.util.concurrent.LinkedBlockingQueue

class ResourceDepository(private val dispatchPeriod: Int) {
    private val resourceQueue: BlockingQueue<Resource> = LinkedBlockingQueue()
    private var timer: Timer? = null
    fun takeResource(): Resource? {
        return try {
            resourceQueue.take()
        } catch (e: InterruptedException) {
            null
        }
    }

    fun addResource(resource: Resource) {
        resourceQueue.add(resource)
    }

    val size: Int
        get() = resourceQueue.size

    fun startResourceMining(
        indX: Int,
        indY: Int,
        devourerStructure: DevourerStructure?,
        startPosition: Position?,
        movingResources: MutableList<Resource?>
    ) {
        if (size > 0) {
            timer = Timer()
            timer!!.schedule(object : TimerTask() {
                override fun run() {
                    val resource = takeResource()
                    if (resource == null) {
                        timer!!.cancel()
                        return
                    }
                    resource.startResource(
                        PositionInd(indX, indY),
                        startPosition,
                        devourerStructure
                    )
                    movingResources.add(resource)
                }
            }, 0, dispatchPeriod.toLong())
        }
    }
}